k9mail/src/main/java/com/fsck/k9/helper/UnreadWidgetProperties.java

```diff
@@ -0,0 +1,140 @@
+        switch (type) {
+            case SEARCH_ACCOUNT:
+            case ACCOUNT:
+                return accountName;
+            case FOLDER:
+                return context.getString(R.string.unread_widget_title, accountName, folderName);
+            default:
+                return null;
+        }
+    }
+
+    public int getUnreadCount(Context context) throws MessagingException {
+        BaseAccount baseAccount = getAccount(context);
+        AccountStats stats;
+        switch (type) {
+            case SEARCH_ACCOUNT:
+                MessagingController controller = MessagingController.getInstance(context);
+                stats = controller.getSearchAccountStatsSynchronous((SearchAccount) baseAccount, null);
+                return stats.unreadMessageCount;
+            case ACCOUNT:
+                Account account = (Account) baseAccount;
+                stats = account.getStats(context);
+                return stats.unreadMessageCount;
+            case FOLDER:
+                return ((Account) baseAccount).getFolderUnreadCount(context, folderName);
+            default:
+                return -1;
+        }
+    }
+
+    public Intent getClickIntent(Context context) {
+        BaseAccount baseAccount = getAccount(context);
+        AccountStats stats;
+        switch (type) {
+            case SEARCH_ACCOUNT:
+                SearchAccount searchAccount = (SearchAccount) baseAccount;
+                return MessageList.intentDisplaySearch(context,
+                        searchAccount.getRelatedSearch(), false, true, true);
+            case ACCOUNT:
+                return getClickIntentForAccount(context);
+            case FOLDER:
+                return getClickIntentForFolder(context);
+            default:
+                return null;
+        }
+    }
+
+    public int getAppWidgetId() {
+        return appWidgetId;
+    }
+
+    public String getAccountUuid() {
+        return accountUuid;
+    }
+
+    public String getFolderName() {
+        return folderName;
+    }
+
+    private void calculateType() {
+        if (SearchAccount.UNIFIED_INBOX.equals(accountUuid) ||
+                SearchAccount.ALL_MESSAGES.equals(accountUuid)) {
+            type = Type.SEARCH_ACCOUNT;
+        } else if(folderName != null) {
+            type = Type.FOLDER;
+        } else {
+            type = Type.ACCOUNT;
+        }
+    }
+
+    private BaseAccount getAccount(Context context) {
+        if (SearchAccount.UNIFIED_INBOX.equals(accountUuid)) {
+            return SearchAccount.createUnifiedInboxAccount(context);
+        } else if (SearchAccount.ALL_MESSAGES.equals(accountUuid)) {
+            return SearchAccount.createAllMessagesAccount(context);
+        }
+        return Preferences.getPreferences(context).getAccount(accountUuid);
+    }
+
+    private Intent getClickIntentForAccount(Context context) {
+        Account account = (Account) getAccount(context);
```

Comment:
At this point we know it's a regular account. Better use `Preferences.getAccount()` directly to avoid the unnecessary work in this class's `getAccount()` and the unsafe type cast. Same in the method below.